### PR TITLE
Permit all

### PR DESCRIPTION
**Pull request by blcham**

### CODE REVIEW

**Comment by kostobog** [comment]
@blcham 
`@PreAuthorize("permitAll()")` it allows unauthorized requests. However, the implementation throws exception as requires a user but there is non.

File: src/main/java/cz/cvut/kbss/study/service/PublishRecordsService.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        this.objectMapper = WebAppConfig.createJsonObjectMapper();
+        this.securityUtils = securityUtils;
+        this.restTemplate = restTemplate;
+        this.configReader = configReader;
+    }
+
+    public RecordImportResult publishRecords(RecordFilterParams filters, Pageable pageSpec){
+        String onPublishRecordsServiceUrl = configReader.getConfig(ConfigParam.ON_PUBLISH_RECORDS_SERVICE_URL);
+        if(onPublishRecordsServiceUrl == null || onPublishRecordsServiceUrl.isBlank()) {
+            LOG.info("No publish service url provided, noop.");
```
Comment (blcham) [inline src/main/java/cz/cvut/kbss/study/service/PublishRecordsService.java:51]: should be at least WARN, provided --> configured

File: src/main/java/cz/cvut/kbss/study/rest/PatientRecordController.java [hunk 2]
```diff
@@ -16,17 +15,14 @@
 import cz.cvut.kbss.study.rest.util.RecordFilterMapper;
 import cz.cvut.kbss.study.rest.util.RestUtils;
 import cz.cvut.kbss.study.security.SecurityConstants;
-import cz.cvut.kbss.study.service.ConfigReader;
-import cz.cvut.kbss.study.service.ExcelRecordConverter;
-import cz.cvut.kbss.study.service.PatientRecordService;
-import cz.cvut.kbss.study.service.UserService;
+import cz.cvut.kbss.study.service.*;
+import cz.cvut.kbss.study.service.security.SecurityUtils;
```
Comment (blcham) [inline src/main/java/cz/cvut/kbss/study/rest/PatientRecordController.java:19]: This does not seem to be used.

File: src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java [hunk 3]
```diff
@@ -128,4 +140,46 @@ public boolean areFromSameInstitution(String username) {
...
+        // TODO - The exchanged token does not contain an IDP field, so the supplier won't know the original identity provider
+        String accessToken = getCurrentToken();
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+        MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+        map.add("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange");
+        map.add("subject_token", accessToken);
+        map.add("subject_token_type", "urn:ietf:params:oauth:token-type:jwt"); // specify that the
+//        map.add("subject_issuer", "vpaf-keycloak-oidc"); // TODO - check if needed
+//        map.add("subject_issuer", "https://kbss.felk.cvut.cz/vietnam-peoples-air-force/services/auth/realms/record-manager");
```
[the remark this answers is not in the document]
Comment (blcham) [reply src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java]: done

File: src/main/java/cz/cvut/kbss/study/rest/PatientRecordController.java [hunk 4]
```diff
@@ -57,20 +53,24 @@ public class PatientRecordController extends BaseController {
...
-                                   UserService userService)  {
+                                   UserService userService, SecurityUtils securityUtils, PublishRecordsService publishRecordsService)  {
         this.recordService = recordService;
         this.eventPublisher = eventPublisher;
         this.excelRecordConverter = excelRecordConverter;
         this.restTemplate = restTemplate;
         this.configReader = configReader;
         this.objectMapper = objectMapper;
         this.userService = userService;
+        this.securityUtils = securityUtils;
```
[the remark this answers is not in the document]
Comment (blcham) [reply src/main/java/cz/cvut/kbss/study/rest/PatientRecordController.java]: done

File: src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java [hunk 5]
```diff
@@ -128,4 +140,46 @@ public boolean areFromSameInstitution(String username) {
...
+        map.add("client_id", "record-manager-server"); // The client_id of the publishing service, i.e. the suppliers record-manager-server
+        map.add("client_secret", config.getConfig(ConfigParam.PUBLISH_RECORDS_SERVICE_SECRET)); // the client secret for the client_id
+//        map.add("audience", "record-manager-server"); // TODO - check if needed, without it the aud field in the token does not contain record-manager-server
+
+        String exchangedToken = null;
+        ResponseEntity<Map> response = null;
+        try {
+            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(map, headers);
+            response = restTemplate.postForEntity(config.getConfig(ConfigParam.EXCHANGE_TOKEN_SERVICE_URL), request, Map.class);
+//            response = restTemplate.postForEntity("https://kbss.felk.cvut.cz/ava/services/auth/realms/record-manager/protocol/openid-connect/token", request, Map.class);
```
[the remark this answers is not in the document]
Comment (blcham) [reply src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java]: done

File: src/main/java/cz/cvut/kbss/study/service/PublishRecordsService.java (hunk 1 above)
Comment (blcham) [reply src/main/java/cz/cvut/kbss/study/service/PublishRecordsService.java]: resolved.

File: src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (blcham) [reply src/main/java/cz/cvut/kbss/study/service/security/SecurityUtils.java]: resolved.

**Pull request merged**